### PR TITLE
Add qutebrowser config directory to python path

### PR DESCRIPTION
**Pull request by jgkamat**
This is done so config.py can import other python files in the config directory. For example, config.py can 'import theme' which would load a theme.py.

The previous path is restored at the end of this function, to avoid tainting qutebrowser's path

Feel free to trash this PR if this dosen't seem like a good approach.

See #2969

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/2970)

### CODE REVIEW

**Comment by jgkamat** [comment]
Tests seem to be failing because it seems that `standarddir.config()` is not available in the test environment. Hardcoding my config directory works for me, so I don't think it's related to the actual method being used.

**Comment by The-Compiler** [comment]
FWIW you can use the `config_tmpdir` fixture in a test which gives you a [py.path](http://py.readthedocs.io/en/latest/path.html) object which gets automatically patched in as config dir.

As quickly mentioned in IRC, I think this needs some more work:

- `sys.path` should also be restored when there was an exception
- `sys.modules` probably needs restoring as well, to make sure no modules imported in `config.py` affect importing future modules in qutebrowser
- I've seen some places mention that using `exec` has some caveats related to imports. I'm not sure what it was anymore, but it's also worth some more investigation.

It might be worth reviewing how other projects handle this first, see the links in #2795 and #30 and [this blogpost](http://lucumr.pocoo.org/2011/2/1/exec-in-python/) for details.

All in all, there are probably more corner-cases hiding here than it seems at first, which is why I didn't do this yet, and it might take me some time to review this :wink:

**Comment by jgkamat** [comment]
From what I can tell, it seems like the cleanest solutions would be:

1. using multiprocessing
2. Using the standard python module system, and making config.py into a function we pass things into rather than top-level python code. 

I don't think either of these are fesable at this point, because (1) is a lot of work, and (2) would require changing the python config file format, so I'm going to try to make this as robust as I can. The best resource I could find (which you linked already) is [this blog post](http://happygiraffe.net/blog/2011/04/28/python-configuration/), which I'm following pretty closely. 

I'll add tests for as many things as I can think of and then I think it will be as good as I can make it (given the limitations). Let me know if there's anything else that I'm missing out (since this is pretty deep python stuff)

File: tests/unit/config/test_configfiles.py [hunk 1]
```diff
@@ -135,6 +135,135 @@ def test_yaml_oserror(fake_save_manager, config_tmpdir):
...
+
+        assert len(api.errors) == 1
+        error = api.errors[0]
+        assert error.text == "Unhandled exception"
+        assert isinstance(error.exception, ImportError)
+
+        tblines = error.traceback.strip().splitlines()
+        assert tblines[0] == "Traceback (most recent call last):"
+        assert (tblines[-1] == "ImportError: No module named 'foobar'" or
+                tblines[-1] == "ModuleNotFoundError: No module named 'foobar'")
```
Comment (jgkamat) [inline tests/unit/config/test_configfiles.py]: Since python changed it's importerror syntax, something like this is needed, but feels hacky. Is there a better way to do this? Maybe I should leave this check off and leave it at the `assert isinstance(error.exception, ImportError)`

Comment (The-Compiler) [reply tests/unit/config/test_configfiles.py]: Or maybe just `assert tblines[-1].endswith("Error: No module named 'foobar'")` or so?

(haven't looked at the rest yet, busy with other stuff still)

File: qutebrowser/config/configfiles.py [hunk 2]
```diff
@@ -198,6 +199,13 @@ def read_config_py(filename=None):
         if not os.path.exists(filename):
             return api
 
+    # Add config directory to python path, so config.py can import other files
+    # in logical places
+    old_state = _pre_config_save()
+    config_dir = os.path.dirname(filename)
+    if config_dir not in sys.path:
+        sys.path = [config_dir] + sys.path
+
```
Comment (The-Compiler) [inline qutebrowser/config/configfiles.py]: I think it'd be better to do this as late as possible, i.e. right before the `exec`.

File: qutebrowser/config/configfiles.py [hunk 3]
```diff
@@ -232,8 +254,17 @@ def read_config_py(filename=None):
             "Unhandled exception",
             exception=e, traceback=traceback.format_exc()))
 
-    api.finalize()
-    return api
+
+def _pre_config_save():
```
Comment (The-Compiler) [inline qutebrowser/config/configfiles.py]: I think this could look much nicer when you make it a context manager:

```python
import contextlib

...

@contextlib.contextmanager
def add_to_sys_path(path):
    # save / change
    yield
    # restore
```

Then:

```python
with add_to_sys_path(config_dir):
    # do config stuff
```

It nicely separates things (even setting sys.path) from running the config, it has setup/cleanup nicer together, and it automatically takes care of doing the cleanup on exceptions as well.

File: qutebrowser/config/configfiles.py [hunk 4]
```diff
@@ -232,8 +254,17 @@ def read_config_py(filename=None):
...
+    old_path = sys.path
+    old_modules = sys.modules.copy()
+    return (old_path, old_modules)
+
+
+def _post_config_load(save_tuple):
+    sys.path = save_tuple[0]
+    for module in set(sys.modules).difference(save_tuple[1]):
+        del sys.modules[module]
```
Comment (The-Compiler) [inline qutebrowser/config/configfiles.py]: Do you do this like this because re-assigning `sys.modules` doesn't work? If so, you might get around it by doing `sys.path[:] = saved_sys_path`, which mutates the existing list.

File: qutebrowser/config/configfiles.py [hunk 5]
```diff
@@ -232,8 +254,17 @@ def read_config_py(filename=None):
...
-    return api
+
+def _pre_config_save():
+    old_path = sys.path
+    old_modules = sys.modules.copy()
+    return (old_path, old_modules)
+
+
+def _post_config_load(save_tuple):
+    sys.path = save_tuple[0]
```
Comment (The-Compiler) [inline qutebrowser/config/configfiles.py]: `sys.path, old_sys_modules = save_tuple` would probably read nicer.

File: tests/unit/config/test_configfiles.py [hunk 6]
```diff
@@ -134,6 +134,144 @@ def test_yaml_oserror(fake_save_manager, config_tmpdir):
     assert error.traceback is None
 
 
+class TestConfigPyModules:
+
+    """Test for ConfigPy Modules."""
```
Comment (The-Compiler) [inline tests/unit/config/test_configfiles.py]: You can remove this docstring - I made them optional for tests, because often they look like this, and say the same thing class names say already.

File: tests/unit/config/test_configfiles.py [hunk 7]
```diff
@@ -134,6 +134,144 @@ def test_yaml_oserror(fake_save_manager, config_tmpdir):
...
+
+        def __init__(self, tmpdir):
+            self._confpy = tmpdir / 'config.py'
+            self.filename = str(self._confpy)
+
+        def write(self, *lines):
+            text = '\n'.join(lines)
+            self._confpy.write_text(text, 'utf-8', ensure=True)
+
+    class QbModulePy:
```
Comment (The-Compiler) [inline tests/unit/config/test_configfiles.py]: I don't think it makes sense to make this a separate class - why not simply adjust `ConfPy` to make it possible to adjust the filename?

File: tests/unit/config/test_configfiles.py [hunk 8]
```diff
@@ -134,6 +134,144 @@ def test_yaml_oserror(fake_save_manager, config_tmpdir):
...
+        return self.QbModulePy(tmpdir)
+
+    def setup_method(self, method):
+        # If we plan to add tests that modify modules themselves, that should
+        # be saved as well
+        TestConfigPyModules._old_sys_path = sys.path.copy()
+
+    def teardown_method(self, method):
+        # Restore path to save the rest of the tests
+        sys.path = TestConfigPyModules._old_sys_path
```
Comment (The-Compiler) [inline tests/unit/config/test_configfiles.py]: Use a fixture instead of `setup_method` / `teardown_method`:

```python
@pytest.fixture(autouse=True)
def restore_sys_path(self):
    old_path = sys.path.copy()
    yield
    sys.path = old_path
```

File: tests/unit/config/test_configfiles.py [hunk 9]
```diff
@@ -134,6 +134,144 @@ def test_yaml_oserror(fake_save_manager, config_tmpdir):
     assert error.traceback is None
 
 
+class TestConfigPyModules:
+
+    """Test for ConfigPy Modules."""
+
+    pytestmark = pytest.mark.usefixtures('config_stub', 'key_config_stub')
+    _old_sys_path = sys.path.copy()
```
Comment (The-Compiler) [inline tests/unit/config/test_configfiles.py]: Is this necessary with the fixture / `setup_method`?

File: tests/unit/config/test_configfiles.py [hunk 10]
```diff
@@ -134,6 +134,144 @@ def test_yaml_oserror(fake_save_manager, config_tmpdir):
...
+        # be saved as well
+        TestConfigPyModules._old_sys_path = sys.path.copy()
+
+    def teardown_method(self, method):
+        # Restore path to save the rest of the tests
+        sys.path = TestConfigPyModules._old_sys_path
+
+    def test_bind_in_module(self, confpy, qbmodulepy):
+        qbmodulepy.write("""def run(config):
+    config.bind(",a", "message-info foo", mode="normal")""")
```
Comment (The-Compiler) [inline tests/unit/config/test_configfiles.py]: You can pass multiple lines to `qbmodulepy`, which'd make this easier to read, IMHO:

```python
        qbmodulepy.write('def run(config):',
                         '    config.bind(",a", "message-info foo", mode="normal")')
```

(same for other tests)

File: tests/unit/config/test_configfiles.py [hunk 11]
```diff
@@ -134,6 +134,144 @@ def test_yaml_oserror(fake_save_manager, config_tmpdir):
...
+        confpy.write("""import qbmodule
+qbmodule.run(config)""")
+        api = configfiles.read_config_py(confpy.filename)
```
Comment (The-Compiler) [inline tests/unit/config/test_configfiles.py]: I added `confpy.read()` in master which does that and checks `api.errors` automatically.

(same for other tests)

File: tests/unit/config/test_configfiles.py [hunk 12]
```diff
@@ -134,6 +134,144 @@ def test_yaml_oserror(fake_save_manager, config_tmpdir):
...
+
+    def teardown_method(self, method):
+        # Restore path to save the rest of the tests
+        sys.path = TestConfigPyModules._old_sys_path
+
+    def test_bind_in_module(self, confpy, qbmodulepy):
+        qbmodulepy.write("""def run(config):
+    config.bind(",a", "message-info foo", mode="normal")""")
+        confpy.write("""import qbmodule
+qbmodule.run(config)""")
```
Comment (The-Compiler) [inline tests/unit/config/test_configfiles.py]: Since you use this a lot, maybe just add a `confpy.write_qbmodule()` which writes this text?

File: tests/unit/config/test_configfiles.py [hunk 13]
```diff
@@ -134,6 +134,144 @@ def test_yaml_oserror(fake_save_manager, config_tmpdir):
...
+        assert tmpdir not in sys.path
+
+    def test_clear_modules(self, confpy, qbmodulepy):
+        qbmodulepy.write("""def run(config):
+    config.bind(",a", "message-info foo", mode="normal")""")
+        confpy.write("""import qbmodule
+qbmodule.run(config)""")
+        api = configfiles.read_config_py(confpy.filename)
+        assert len(api.errors) == 0
+        assert "qbmodule" not in sys.modules.keys()
```
Comment (The-Compiler) [inline tests/unit/config/test_configfiles.py]: `test_clear_path` and `test_clear_modules` are really similar to `test_bind_in_module` - I know some people prefer tests to be strictly separated, but I think it'd make more sense to just check `sys.path` and `sys.modules` in `test_bind_in_module` since it's really the same "concept" - I'll leave the decision to you though.

File: tests/unit/config/test_configfiles.py [hunk 14]
```diff
@@ -134,6 +134,144 @@ def test_yaml_oserror(fake_save_manager, config_tmpdir):
...
+        assert len(api.errors) == 1
+        error = api.errors[0]
+        assert error.text == "Unhandled exception"
+        assert isinstance(error.exception, ZeroDivisionError)
+
+        tblines = error.traceback.strip().splitlines()
+        assert tblines[0] == "Traceback (most recent call last):"
+        assert tblines[-1] == "ZeroDivisionError: division by zero"
+        assert "    1/0" in tblines
+        assert tmpdir not in sys.path
```
Comment (The-Compiler) [inline tests/unit/config/test_configfiles.py]: Like above, I think these two tests could be combined.

File: tests/unit/config/test_configfiles.py [hunk 15]
```diff
@@ -134,6 +134,144 @@ def test_yaml_oserror(fake_save_manager, config_tmpdir):
...
+
+        assert len(api.errors) == 1
+        error = api.errors[0]
+        assert error.text == "Unhandled exception"
+        assert isinstance(error.exception, ZeroDivisionError)
+
+        tblines = error.traceback.strip().splitlines()
+        assert tblines[0] == "Traceback (most recent call last):"
+        assert tblines[-1] == "ZeroDivisionError: division by zero"
+        assert "    1/0" in tblines
```
Comment (The-Compiler) [inline tests/unit/config/test_configfiles.py]: I don't think it's necessary to test the traceback in those tests again, they are focused on a different thing and we already have a test looking at the tracebacks.

File: tests/unit/config/test_configfiles.py [hunk 16]
```diff
@@ -134,6 +134,144 @@ def test_yaml_oserror(fake_save_manager, config_tmpdir):
...
+        assert isinstance(error.exception, ImportError)
+
+        tblines = error.traceback.strip().splitlines()
+        assert tblines[0] == "Traceback (most recent call last):"
+        assert tblines[-1].endswith("Error: No module named 'foobar'")
+
+    def test_no_double_if_path_exists(self, confpy, qbmodulepy, tmpdir):
+        sys.path.insert(0, tmpdir)
+        confpy.write("""import sys
+if sys.path[1:].count(sys.path[0]) != 0:
```
Comment (The-Compiler) [inline tests/unit/config/test_configfiles.py]: `if sys.path[0] in sys.path[1:]:`

File: qutebrowser/config/configfiles.py (hunk 4 above)
Comment (jgkamat) [reply qutebrowser/config/configfiles.py]: Yup, re-assigning sys.modules didn't seem to work.

sys.path is a dictionary, so that method dosen't work (`TypeError: unhashable type: 'slice'`).

If you have other ideas, they would be nice, because this is a hack...

File: tests/unit/config/test_configfiles.py (hunk 13 above)
Comment (jgkamat) [reply tests/unit/config/test_configfiles.py]: This  reduces a ton of duplication, so I agree with you :)

File: tests/unit/config/test_configfiles.py (hunk 8 above)
Comment (jgkamat) [reply tests/unit/config/test_configfiles.py]: yup, this is a lot cleaner. I probably should learn how to actually use pytest properly...

File: qutebrowser/config/configfiles.py [hunk 17]
```diff
@@ -222,20 +223,43 @@ def read_config_py(filename=None):
         raise configexc.ConfigFileErrors(basename, [desc])
     except SyntaxError as e:
         desc = configexc.ConfigErrorDesc("Syntax Error", e,
-                                         traceback=traceback.format_exc())
+                                        traceback=traceback.format_exc())
```
Comment (The-Compiler) [inline qutebrowser/config/configfiles.py]: That was probably unintended?

File: qutebrowser/config/configfiles.py [hunk 18]
```diff
@@ -222,20 +223,43 @@ def read_config_py(filename=None):
...
 
     try:
-        exec(code, module.__dict__)
+        # Save and restore sys variables
+        with saved_sys_properties():
+            # Add config directory to python path, so config.py can import
+            # other files in logical places
+            config_dir = os.path.dirname(filename)
+            if config_dir not in sys.path:
+                sys.path = [config_dir] + sys.path
```
Comment (The-Compiler) [inline qutebrowser/config/configfiles.py]: `sys.path.insert(0, config_dir)` maybe?

File: qutebrowser/config/configfiles.py [hunk 19]
```diff
@@ -222,20 +223,43 @@ def read_config_py(filename=None):
...
+
+            exec(code, module.__dict__)
     except Exception as e:
         api.errors.append(configexc.ConfigErrorDesc(
             "Unhandled exception",
             exception=e, traceback=traceback.format_exc()))
-
```
Comment (The-Compiler) [inline qutebrowser/config/configfiles.py]: That was probably unintended?

File: qutebrowser/config/configfiles.py [hunk 20]
```diff
@@ -222,20 +223,43 @@ def read_config_py(filename=None):
...
+@contextlib.contextmanager
+def saved_sys_properties():
+    """Save various sys properties such as sys.path and sys.modules."""
+    old_path = sys.path
+    old_modules = sys.modules.copy()
+
+    try:
+        yield
+    except:
+        raise
```
Comment (The-Compiler) [inline qutebrowser/config/configfiles.py]: No need for the `except: raise`

File: qutebrowser/config/configfiles.py [hunk 21]
```diff
@@ -222,20 +223,43 @@ def read_config_py(filename=None):
...
+    finally:
+        sys.path = old_path
+        for module in set(sys.modules).difference(old_modules):
+            del sys.modules[module]
```
Comment (The-Compiler) [inline qutebrowser/config/configfiles.py]: I thought you could do:

```python
sys.modules.clear()
sys.modules.update(old_modules)
```

but Python doesn't like this:

```
>>> sys.modules.clear()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: lost builtins module
```

Comment (The-Compiler) [reply qutebrowser/config/configfiles.py]: Also, looking at my links again, I'm not sure anymore whether restoring `sys.modules` will actually solve or introduce more edge cases 😆 I *thought* I've seen this in some code I did read, but now I can't find it anymore... Did you see it somewhere?

File: tests/unit/config/test_configfiles.py [hunk 22]
```diff
@@ -134,6 +134,72 @@ def test_yaml_oserror(fake_save_manager, config_tmpdir):
     assert error.traceback is None
 
 
+class TestConfigPyModules:
+
+    pytestmark = pytest.mark.usefixtures('config_stub', 'key_config_stub')
+
+    @pytest.fixture
+    def confpy(self, tmpdir):
+        return TestConfigPy.ConfPy(tmpdir)
```
Comment (The-Compiler) [inline tests/unit/config/test_configfiles.py]: Just make the `ConfPy` class module-scoped at that point :wink:

File: tests/unit/config/test_configfiles.py [hunk 23]
```diff
@@ -144,8 +210,8 @@ class ConfPy:
 
         """Helper class to get a confpy fixture."""
 
-        def __init__(self, tmpdir):
-            self._confpy = tmpdir / 'config.py'
+        def __init__(self, tmpdir, filename: str = "config.py"):
+            self._confpy = tmpdir / filename
```
Comment (The-Compiler) [inline tests/unit/config/test_configfiles.py]: nitpick: Maybe call this `self._file` or so now?

File: qutebrowser/config/configfiles.py (hunk 18 above)
Comment (jgkamat) [reply qutebrowser/config/configfiles.py]: I can either make a copy (via `sys.path.copy()` or similar) when save/restoring, or I can avoid touching this variable inside the `when`. I think that the copy is better, since then whoever is using the contextmanager dosen't need to worry about modifying the original list, so I'll stick to that.

File: qutebrowser/config/configfiles.py (hunk 21 above)
Comment (jgkamat) [reply qutebrowser/config/configfiles.py]: save/restoring of `sys.modules` is kind of a hack (since it dosen't properly clean up the modules). That blog post says that there's no better way to do this other than multiprocessing though, so I don't know if there's a better way (other than using `import` rather than `exec`).

File: qutebrowser/config/configfiles.py (hunk 18 above)
Comment (The-Compiler) [reply qutebrowser/config/configfiles.py]: I don't understand your comment - wouldn't `sys.path = [config_dir] + sys.path` and `sys.path.insert(0, config_dir)` do the same here?

**Comment by The-Compiler** [comment]
re `multiprocessing`: A cleaner way with the same effect would be to start a Python subprocess ourselves which implements the config API based on a serialized config it gets passed. Then that config API and `config.py` would run in the subprocess, and the subprocess would then serialize the changed config again for the main process.

I'm not convinced it's worth the effort though. I think it's better to just try this and see how things go.

File: tests/unit/config/test_configfiles.py [hunk 24]
```diff
@@ -134,32 +134,103 @@ def test_yaml_oserror(fake_save_manager, config_tmpdir):
...
+    def confpy(self, tmpdir):
+        return ConfPy(tmpdir)
+
+    @pytest.fixture
+    def qbmodulepy(self, tmpdir):
+        return ConfPy(tmpdir, filename="qbmodule.py")
+
+    @pytest.fixture(autouse=True)
+    def restore_sys_path(self):
+        old_path = sys.path.copy()
```
Comment (The-Compiler) [inline tests/unit/config/test_configfiles.py]: Oh, missing `sys` import.

Comment (jgkamat) [reply tests/unit/config/test_configfiles.py]: hmm, it's weird this is passing on my machine but not on travis (there is a sys import on the top of the file). I'll try adding a import sys in the fixure (are fixures isolated like that)?

Comment (The-Compiler) [reply tests/unit/config/test_configfiles.py]: Ah, I think I know what's happening: I removed the import in ef1c83862b8166066a860fbfac93a19ec26bb8c3 and Travis tests it against the latest master, while your local branch is on an older master. So the right solution is to rebase your local branch on the latest master, remove this, and re-add it at the top :wink:

**Comment by The-Compiler** [comment]
LGTM now, will merge once the tests passed.

@jgkamat If you see a pylint failure about an unused `configtypes` import in `test_runners`, don't worry, I just [pushed a fix](https://github.com/qutebrowser/qutebrowser/commit/10016ae240863d68949681f789a84d129698b7a1) to master.

@rcorre Sorry for all the mails by the way - looks like you've been pulled in as a reviewer on every PR where someone accidentally rebased on the wrong branch, as then their PR touched some of the completion files.

**Comment by The-Compiler** [comment]
Thanks! I also added some docs in ebf378a945f99eec7e07e96e6f593c1e16949b87.

**Comment by rcorre** [comment]
@The-Compiler no problem, I actually would have given it a look if I had the time,  but I've been moving the past week so I'm barely keeping up with my own PR's >.<
